### PR TITLE
Implement UDS service with example

### PR DESCRIPTION
**Pull request by Jhynjhiruu**
This PR adds basic local networking functionality with the `UDS` service.
It's very much not feature-complete, but enough is implemented to be able to interface with the devkitPRO local networking example - i.e. joining a network, creating a network, sending and pulling packets, and basic configuration. This is probably enough functionality for most homebrew applications, and I plan to try and implement a Download Play sniffer with it.

Todo: finish implementing the rest of the functions; test properly; cleanup code?

An example is included (`local-networking.rs`; this is a mostly-but-not-entirely-complete port of the devkitPRO local networking example).

### CODE REVIEW

File: ctru-rs/src/services/ir_user.rs
Comment (Meziu) [inline ctru-rs/src/services/ir_user.rs:153]: @AzureMarker Ehm, do you remember why a closure is used here? It doesn't seem to be necessary.

File: ctru-rs/src/services/uds.rs
Comment (Meziu) [inline ctru-rs/src/services/uds.rs]: Unknown members, padding bytes and other "generic" data for which use isn't clear shouldn't be `pub`.

File: ctru-rs/src/services/uds.rs
Comment (Meziu) [inline ctru-rs/src/services/uds.rs:54]: Nit: This currently doesn't serve a purpose. If possible, I'd like to keep the Error types distinct (though, it might become useful if this module were to be expanded upon).

File: ctru-rs/src/services/uds.rs
Comment (Meziu) [inline ctru-rs/src/services/uds.rs]: This should probably be a developer comment, rather than documentation.

File: ctru-rs/src/services/uds.rs
Comment (Meziu) [inline ctru-rs/src/services/uds.rs]: Same as `NetworkScanInfo`.

File: ctru-rs/src/services/uds.rs
Comment (Meziu) [inline ctru-rs/src/services/uds.rs]: It would be nice to add descriptions of the members.

File: ctru-rs/src/services/uds.rs
Comment (Meziu) [inline ctru-rs/src/services/uds.rs:417]: This is quite a large number. It's more than 16kB of allocated memory for *each* call of the `scan()` function. Do we need this much memory?

File: ctru-rs/src/services/uds.rs
Comment (Meziu) [inline ctru-rs/src/services/uds.rs]: Using `CString::new(username)` is preferred.

File: ctru-rs/src/services/uds.rs
Comment (Meziu) [inline ctru-rs/src/services/uds.rs:478]: What happens if the pointer passed is `NUL`?

File: ctru-rs/src/services/uds.rs
Comment (Meziu) [inline ctru-rs/src/services/uds.rs:583]: Nit: here you could use `MaybeUninit`, to tie in better with the rest of the code.

File: ctru-rs/src/services/uds.rs
Comment (Meziu) [inline ctru-rs/src/services/uds.rs]: I believe so, the internal `ctru_sys::udsConnectionStatus` contains informations specific to an existing network.

File: ctru-rs/src/services/uds.rs
Comment (Meziu) [inline ctru-rs/src/services/uds.rs]: `ctru_sys::udsConnectionStatus` should get wrapped within a more standard struct that holds a safe method API.

File: ctru-rs/src/services/uds.rs
Comment (Meziu) [inline ctru-rs/src/services/uds.rs]: `ctru_sys::UDS_BROADCAST_NETWORKNODEID` and `ctru_sys::UDS_HOST_NETWORKNODEID` should be handled as bitflags. Programmatically adding nodes to the flag would be quite useful as well.

File: ctru-rs/src/services/uds.rs
Comment (Meziu) [inline ctru-rs/src/services/uds.rs]: It might be wiser to `truncate()` and possibly `shrink_to_fit()` the original vector, rather than reallocating.

File: ctru-rs/src/services/uds.rs
Comment (Meziu) [inline ctru-rs/src/services/uds.rs:1173]: Is the `appdata` saved within the struct? If it just holds the pointer it might cause segmentation faults when the original buffer is dropped.

**Review by Meziu — COMMENTED** [review]
Great PR! A bit long for my liking, but I saw you followed our standards for the public API. Sadly, I have no way to thoroughly test the changes since I can't make a local network (I only have one 3DS). Is there any way I can test things out, maybe with an emulator?

File: ctru-rs/src/services/ir_user.rs
Comment (AzureMarker) [reply ctru-rs/src/services/ir_user.rs:153]: This was done so I could use `?`, like the unstable `try` block.

**Comment by Jhynjhiruu** [comment]
> Is there any way I can test things out, maybe with an emulator?

I've looked a little bit into it and I think it's maybe possible to connect to a 3DS's network with another device, but it'd require more knowledge about how Wi-Fi works than I have.
I think Citra supports relaying local networking data for multiplayer, but I don't know how to set it up.

File: ctru-rs/src/services/ir_user.rs
Comment (Jhynjhiruu) [reply ctru-rs/src/services/ir_user.rs:153]: This change was the easiest way to fix a compile error introduced by my implementation of `FromResidual`. If I handle errors a different way, this change can be reverted.

File: ctru-rs/src/services/uds.rs
Comment (Jhynjhiruu) [reply ctru-rs/src/services/uds.rs]: Right, sorry. I basically blindly copied that without really thinking about it.

File: ctru-rs/src/services/uds.rs
Comment (Jhynjhiruu) [reply ctru-rs/src/services/uds.rs:54]: There are a bunch of functions which return this error type that also need to be able to return ctru-rs errors. I'm not massively well-versed in how ctru-rs's error system works, and this was the solution I came up with for this. If there's a better way to handle it, I'm very happy to change it, but I don't know how to do that myself.

File: ctru-rs/src/services/uds.rs
Comment (Jhynjhiruu) [reply ctru-rs/src/services/uds.rs:417]: This buffer has to hold all of the info for beacons found during the scanning process, which can essentially be arbitrarily large depending on how many beacons there are; the devkitPRO example uses 16KB.
However, looking at the example code, I realise now I was a bit silly, because the buffer can absolutely just be allocated on the heap once when the service is initialised and reused for each scan. Is this acceptable?

File: ctru-rs/src/services/uds.rs
Comment (Jhynjhiruu) [reply ctru-rs/src/services/uds.rs]: How so? `CString::new` can't take an `Option`, and also we want to raise an error if the provided username contains NULL bytes, so that the user doesn't unexpectedly set a different username to what they thought.

File: ctru-rs/src/services/uds.rs
Comment (Jhynjhiruu) [reply ctru-rs/src/services/uds.rs:478]: The 3DS's system username is used.

File: ctru-rs/src/services/uds.rs
Comment (Meziu) [reply ctru-rs/src/services/uds.rs:417]: I guess so, it would surely help mitigate memory fragmentation. Now I imagine a room filled with 200 Nintendo 3DS trying to download play and the chaos that would ensue 🥸

File: ctru-rs/src/services/uds.rs
Comment (Jhynjhiruu) [reply ctru-rs/src/services/uds.rs:583]: Does `MaybeUninit` work with non-constant sizes? I _could_ always allocate a max-sized array, but that seems a bit wasteful.

File: ctru-rs/src/services/uds.rs
Comment (Jhynjhiruu) [reply ctru-rs/src/services/uds.rs]: I haven't yet checked for sure, but I'm pretty sure that network node IDs aren't actually bitflags, even though they probably should be - the devkitPRO example implies that you access nodes numerically. I'll check this with my consoles.

File: ctru-rs/src/services/uds.rs
Comment (Jhynjhiruu) [reply ctru-rs/src/services/uds.rs]: I forgot to check the standard library documentation to see if there's a better way to do this. Whoops.

File: ctru-rs/src/services/uds.rs
Comment (Jhynjhiruu) [reply ctru-rs/src/services/uds.rs:1173]: It's not entirely clear from the `libctru` documentation, but reading [the code](https://github.com/devkitPro/libctru/blob/master/libctru/source/services/uds.c#L744), it seems like it's copied to a TLS buffer. Oh god, not TLS again.

File: ctru-rs/src/services/uds.rs
Comment (Meziu) [reply ctru-rs/src/services/uds.rs:478]: You should specify that in the docs 👍

File: ctru-rs/src/services/uds.rs
Comment (Jhynjhiruu) [reply ctru-rs/src/services/uds.rs]: Yep, confirmed - ID 0x0001 is the host, then 0x0002, 0x0003, 0x0004 etc. are the nodes on the network. Sending to node 0x0000 does nothing, sending to node 0xFFFF broadcasts to all nodes, sending to nodes not on the network does nothing, sending to nodes > 16 also does nothing, and sending to the device's own node returns an error (0xC90113FA, uds, status, wrong arg, not found). I suppose I could add an enum for this?

File: ctru-rs/src/services/uds.rs
Comment (Jhynjhiruu) [reply ctru-rs/src/services/uds.rs]: It seems like calling it when not connected actually returns success. On New 3DS, I get
```rs
ConnectionStatus {
    status: 3,
    unk_x4: 0,
    cur_node_id: 2,
    unk_xa: 0,
    unk_xc: [0, 0, 0, 0, 0, 0, 0, 0],
    total_nodes: 0,
    max_nodes: 0,
    node_bitmask: 0,
}
```
and on Old 3DS I get
```rs
ConnectionStatus {
    status: 3,
    unk_x4: 0,
    cur_node_id: 0,
    unk_xa: 0,
    unk_xc: [0, 0, 0, 0, 0, 0, 0, 0],
    total_nodes: 0,
    max_nodes: 0,
    node_bitmask: 0,
}
```
consistently (they're identical besides `cur_node_id`).

Comment (Meziu) [reply ctru-rs/src/services/uds.rs]: Interesting. At this point I wonder whether the `status` member itself specifies whether or not the client is connected to a network. We should probably try to understand its meaning.

Comment (Jhynjhiruu) [reply ctru-rs/src/services/uds.rs]: `libctru` doesn't describe it at all, so it would very much be a reverse-engineering effort. From my 20 seconds of experimentation, it looks like a value of 6 represents connected. Presumably it's a bitfield.

File: ctru-rs/src/services/uds.rs
Comment (Meziu) [reply ctru-rs/src/services/uds.rs]: I think I wanted to remove a degree of complexity (`map` -> `new` with separate `match`) and convert the final error directly, but after revisiting the changes with the other parts of the code it's definitely fine as is.

File: ctru-rs/src/services/uds.rs
Comment (Meziu) [reply ctru-rs/src/services/uds.rs:583]: I got confused with uninitialized slices, sorry :sweat_smile:

File: ctru-rs/src/services/uds.rs
Comment (Meziu) [reply ctru-rs/src/services/uds.rs]: So the packet needs to be sent out each time for every node that needs to receive it? Other than for broadcasts I guess. Wack,  I just believed it worked like/together with the nodes bitmask.

File: ctru-rs/src/services/uds.rs
Comment (Meziu) [reply ctru-rs/src/services/uds.rs:1173]: I mean, it's not like TLS storage doesn't work. The linker just hates us in particular. :laughing:

**Comment by Jhynjhiruu** [comment]
Functions left to implement:

- [ ] udsGenerateNodeInfo (not really intended for use standalone, I think, and doesn't really need to be in the safe API)
- [ ] udsGetNodeInfoUsername (the safe API replaces the `NodeInfo` type and that decodes the string, so there's no need for this)
- [ ] udsCheckNodeInfoInitialized (I have no idea why this is supposed to be useful, since you always know this already)
- [ ] udsGenerateDefaultNetworkStruct (also not really intended for use standalone, used as part of `Uds::new()`)
- [x] udsWaitDataAvailable (would be good to have, but not really needed since you can just check whether `Uds::pull_packet()` returns `Some` or `None`)
- [x] udsEjectClient
- [x] udsEjectSpectator
- [ ] udsUpdateNetworkAttribute (definitely not really intended for use standalone, other functions call this internally)
- [x] udsSetNewConnectionsBlocked
- [x] udsAllowSpectators
- [x] udsGetNodeInformation

So, only about 5½ that would be actually useful, and they're mostly trivial.

File: ctru-rs/src/services/uds.rs
Comment (ian-h-chamberlain) [inline ctru-rs/src/services/uds.rs:184]: Personally, I'd probably implement this by just wrapping the underlying type, especially since a lot of the debug output will just end up as padding this way. Maybe something like this?

```rust
#[derive(Copy, Clone)]
pub struct NodeInfo(ctru_sys::udsNodeInfo);

impl Debug for NodeInfo {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "NodeInfo(")?;

        let user_and_flag = unsafe { self.0.__bindgen_anon_1.__bindgen_anon_1 };

        f.debug_struct("udsNodeInfo")
            .field("uds_friendcodeseed", &self.0.uds_friendcodeseed)
            .field(
                "username",
                &String::from_utf16_lossy(&user_and_flag.username),
            )
            .field("flag", &user_and_flag.flag)
            .field("NetworkNodeID", &self.0.NetworkNodeID)
            .finish()?;

        write!(f, ")")
    }
}
```

File: ctru-rs/src/services/uds.rs
Comment (ian-h-chamberlain) [inline ctru-rs/src/services/uds.rs]: Why does this need to be reimplemented in Rust? Again could we just wrap the upstream type?

File: ctru-rs/src/services/uds.rs
Comment (ian-h-chamberlain) [inline ctru-rs/src/services/uds.rs:146]: Might make sense for us to just generate this constant as a `u16` via bindgen's [`ParseCallbacks::int_macro`](https://docs.rs/bindgen/latest/bindgen/callbacks/trait.ParseCallbacks.html#method.int_macro), to help prevent conversions.

In fact I wonder if we can't do this for a bunch of other constants and stop using `as`, maybe I'll file a separate issue for this more broadly

File: ctru-rs/src/services/uds.rs
Comment (ian-h-chamberlain) [inline ctru-rs/src/services/uds.rs]: I think you could simply this a bit if we had an `impl From<std::ffi::NulError> for Error`:

```suggestion
        let cstr = username.map(CString::new).transpose()?;
```

File: ctru-rs/src/services/uds.rs
Comment (ian-h-chamberlain) [inline ctru-rs/src/services/uds.rs:480]: Does this need to be `Pin` or something, to make sure the string isn't moved while the service is initialized? I would imagine that if you moved the resulting `UDS` struct the pointer copied into the `udsInit` could become invalid.

File: ctru-rs/src/services/uds.rs
Comment (ian-h-chamberlain) [inline ctru-rs/src/services/uds.rs:409]: Is there any reason to make this an associated const if it's not `pub`? The extra comment might be helpful but only if these are exported

File: ctru-rs/src/services/uds.rs
Comment (ian-h-chamberlain) [inline ctru-rs/src/services/uds.rs]: This could probably avoid a hardcode with a little trickery (not sure if `SCAN_BUF_SIZE` and `SHAREDMEM_SIZE` would benefit from something similar): https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=9e32fe75295dc99916039224351f337c

File: ctru-rs/src/services/uds.rs
Comment (ian-h-chamberlain) [inline ctru-rs/src/services/uds.rs]: ?

```suggestion
                .map(NetworkScanInfo::from)
```

File: ctru-rs/src/services/uds.rs
Comment (ian-h-chamberlain) [inline ctru-rs/src/services/uds.rs:686]: Does this always need to be nul-terminated? I Wonder if `&CStr` makes sense here instead of `&[u8]` but not positive what kinds of values might be passed in

File: ctru-rs/src/services/uds.rs
Comment (ian-h-chamberlain) [reply ctru-rs/src/services/uds.rs]: I think I found some details, seems like it's more of a weird enum with sparse value choices? https://www.3dbrew.org/wiki/NWMUDS:GetConnectionStatus#Status_values

File: ctru-rs/src/services/uds.rs
Comment (ian-h-chamberlain) [inline ctru-rs/src/services/uds.rs:964]: What does this status represent? it still seems like it would be an error of some kind but we
return `Ok` for this?

File: ctru-rs/src/services/uds.rs
Comment (Jhynjhiruu) [reply ctru-rs/src/services/uds.rs:686]: No, it can be whatever data you like. Games conventionally use a null-terminated string, but it's just raw bytes.

File: ctru-rs/src/services/uds.rs
Comment (Jhynjhiruu) [reply ctru-rs/src/services/uds.rs:964]: See [libctru](https://github.com/devkitPro/libctru/blob/master/libctru/include/3ds/services/uds.h#L23) and [the UDS example](https://github.com/devkitPro/3ds-examples/blob/master/network/uds/source/uds.c#L286C7-L286C34). There's probably a nicer way to handle this, but this does work.

File: ctru-rs/src/services/uds.rs
Comment (Jhynjhiruu) [reply ctru-rs/src/services/uds.rs]: Uuh, yeah, that'll work probably. Didn't know you could do that!

File: ctru-rs/src/services/uds.rs
Comment (Jhynjhiruu) [reply ctru-rs/src/services/uds.rs:409]: There's no good reason (as far as I'm aware) to have the user able to change them, but they're properties of the data structure, rather than something you'd want to limit within a single function.

File: ctru-rs/src/services/uds.rs
Comment (Jhynjhiruu) [reply ctru-rs/src/services/uds.rs:480]: `udsInit` copies the string (to be precise, it converts it to UTF-16), so I think it's safe.

File: ctru-rs/src/services/uds.rs
Comment (Jhynjhiruu) [reply ctru-rs/src/services/uds.rs]: I'm not happy with most of the error handling in this module, so spending some time reworking it all would be useful I think.

File: ctru-rs/src/services/uds.rs
Comment (Jhynjhiruu) [reply ctru-rs/src/services/uds.rs:146]: Is that likely to get done any time soon? If so, I'll hold off from fixing this until it's merged.

File: ctru-rs/src/services/uds.rs
Comment (Jhynjhiruu) [reply ctru-rs/src/services/uds.rs]: Could do, but then all the accesses to the fields would have to be through methods. I've forgotten why I didn't do it that way originally at this point.

File: ctru-rs/src/services/uds.rs
Comment (Jhynjhiruu) [reply ctru-rs/src/services/uds.rs:184]: Yeah, that's probably better than what I've done.

File: ctru-rs/examples/local-networking.rs
Comment (ian-h-chamberlain) [inline ctru-rs/examples/local-networking.rs:1]: Wow, this is a pretty intense example but it seems necessary given how many features there are in the `uds` service. I was wondering if some of this functionality could go onto `Uds` itself instead of just in this example, but honestly on first read I don't see any obvious things that make sense, so 👍 good with me

File: ctru-rs/src/services/uds.rs
Comment (ian-h-chamberlain) [reply ctru-rs/src/services/uds.rs:480]: ah gotcha, perfect

File: ctru-rs/src/services/uds.rs
Comment (ian-h-chamberlain) [reply ctru-rs/src/services/uds.rs:146]: Probably not, I'll capture this in a separate issue and you can leave this as-is (edit: #166)

File: ctru-rs/src/services/uds.rs
Comment (ian-h-chamberlain) [reply ctru-rs/src/services/uds.rs:964]: Ok, so this error code basically means a packet was still sent but some other error occurred, I guess? Seems fine

Comment (Jhynjhiruu) [reply ctru-rs/src/services/uds.rs:964]: Yeah, something like that. Maybe it'd be a good idea to hardcode it as a constant somewhere rather than calling it in the function, though - `MAKERESULT` presumably isn't `const` so it'd have to be a magic number with some explanation.

Comment (ian-h-chamberlain) [reply ctru-rs/src/services/uds.rs:964]: If it isn't `const`, we probably can/should make it `const`, same for any of those other helper functions that just do bit operations on `Result`

File: ctru-rs/examples/local-networking.rs
Comment (adryzz) [inline ctru-rs/examples/local-networking.rs:43]: This looks a lot like a manually implemented future state machine. could this have worked better with async?

Comment (Jhynjhiruu) [reply ctru-rs/examples/local-networking.rs:43]: Very possibly! I'm scared of `async`.

File: ctru-rs/src/services/uds.rs
Comment (Jhynjhiruu) [reply ctru-rs/src/services/uds.rs:964]: Presumably, this would need to be changed in ctru_sys. Once somebody does that, I'll change this to be a little more elegant.

File: ctru-rs/examples/local-networking.rs
Comment (Jhynjhiruu) [reply ctru-rs/examples/local-networking.rs:43]: I've done some basic reading about futures, and I don't understand how this could be done with async. Could you explain how that would work? It would definitely be nice to clean up the code if it's possible.

**Review by ian-h-chamberlain — APPROVED** [review]

File: ctru-rs/src/services/uds.rs
Comment (Meziu) [reply ctru-rs/src/services/uds.rs]: The `TODO` doc comment should be removed now that it's fixed.

File: ctru-rs/src/services/uds.rs
Comment (Meziu) [inline ctru-rs/src/services/uds.rs]: Nit: remove the `get_*` prefix.

**Review by Meziu — APPROVED** [review]

**Pull request merged**
